cpp/src/arrow/util/async_generator.h

```diff
@@ -235,29 +505,213 @@ class ReadaheadGenerator {
+      bool should_continue = false;
+      {
+        auto guard = state->mutex.Lock();
+        if (!maybe_next.ok() || IterationTraits<AsyncGenerator<T>>::IsEnd(*maybe_next)) {
+          state->source_exhausted = true;
+          if (--state->num_active_subscriptions == 0) {
+            state->finished = true;
+            should_purge = true;
+          }
+        } else {
+          state->active_subscriptions[index] = *maybe_next;
+          should_continue = true;
+        }
+      }
+      if (should_continue) {
+        (*maybe_next)().AddCallback(InnerCallback{state, index});
+      } else if (should_purge) {
+        // At this point state->finished has been marked true so no one else
+        // will be interacting with waiting_jobs and we can iterate outside lock
+        while (!state->waiting_jobs.empty()) {
+          state->waiting_jobs.front()->MarkFinished(IterationTraits<T>::End());
+          state->waiting_jobs.pop_front();
+        }
+      }
+    }
+    std::shared_ptr<State> state;
+    std::size_t index;
+  };
+
+  std::shared_ptr<State> state_;
+};
+
+/// \brief Creates a generator that takes in a stream of generators and pulls from up to
+/// max_subscriptions at a time
 ///
-/// This generator is not async-reentrant
-template <typename T, typename V>
-AsyncGenerator<V> MakeAsyncGenerator(AsyncGenerator<T> generator,
-                                     Transformer<T, V> transformer) {
-  return TransformingGenerator<T, V>(generator, transformer);
+/// Note: This is the equivalent of Rx::MergeMap.  This may deliver items out of
+/// sequence. For example, items from the third AsyncGenerator generated by the source
+/// may be emitted before some items from the first AsyncGenerator generated by the
+/// source.
+///
+/// This generator expects source to be async-reentrant regardless of whether this
+/// generator is async-reentrant or not (unless max_readahead is 1)
+/// This generator will not pull from the individual subscriptions reentrantly.  Add
+/// readahead to the individual subscriptions if that is desired.
+/// This generator is async-reentrant
+///
+/// This generator may queue up to max_readahead instances of T
+template <typename T>
+AsyncGenerator<T> MakeMergeMapGenerator(AsyncGenerator<AsyncGenerator<T>> source,
+                                        int max_readahead) {
+  return MergeMapGenerator<T>(std::move(source), max_readahead);
 }
 
-/// \brief Transfers execution of the generator onto the given executor
+/// \brief Creates a generator that takes in a stream of generators and pulls from each
+/// one in sequence.
+///
+/// Note: This is the equivalent of Rx::ConcatMap.  One could conceivably create the
+/// equivalent of Rx::MergeMap by pulling from `source` before the last emitted
+/// generator is exhausted.  Then ConcatMap folds into the special case of MergeMap with
+/// one max subscription (which is how its handled in Rx)  TODO(ARROW-11800).  Deferring
+/// that now for simplicity.
```

Comment:
It seems this TODO is already done, assuming max_readahead above is really max_subscriptions.

Reply:
Good catch.  I removed the comment.